pageserver/src/layered_repository.rs

```diff
@@ -1428,6 +1436,77 @@ impl LayeredTimeline {
                 continue 'outer;
             }
 
+            // 4. Does this layer serve as a tombstome for some older layer?
+            if l.is_dropped() {
+                let prior_lsn = l.get_start_lsn().checked_sub(1u64).unwrap();
+
+                // Check if this layer serves as a tombstone for this timeline
+                // We have to do this separately from timeline check below,
+                // because LayerMap of this timeline is already locked.
+                let mut is_tombstone = layers.layer_exists_at_lsn(l.get_seg_tag(), prior_lsn);
+                if is_tombstone {
+                    info!(
+                        "earlier layer exists at {} in {}",
+                        prior_lsn, self.timelineid
+                    );
+                }
+                // Now check ancestor timelines, if any
+                else if let Some(ancestor) = &self.ancestor_timeline {
+                    let prior_lsn = ancestor.get_last_record_lsn();
```

Comment:
Could we check the ancestor's layer map directly? Then logic would be the same for the current timeline and the ancestors.

Reply:
Timeline functions `ancestor.get_relish_size()/ ancestor.get_rel_exists()` will recurse down through all parents.
We can repeat this logic in `gc_timeline` and use layer maps directly, but I don't think that it will be more convenient

Reply:
Could you call `ancestor.layers.lock().layer_exists_at_lsn(...)`?

Reply:
> Could you call `ancestor.layers.lock().layer_exists_at_lsn(...)`?

It will only check the nearest ancestor.
And it's possible that the parent of the timeline doesn't have the layer for the segment in question, while the grandparent does. To iterate through the inheritance tree we need to use Timeline functions.